verilog/tools/kythe/kythe_facts_extractor.h

```diff
@@ -0,0 +1,163 @@
+// Copyright 2017-2020 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef VERIBLE_VERILOG_TOOLS_KYTHE_KYTHE_FACTS_EXTRACTOR_H_
+#define VERIBLE_VERILOG_TOOLS_KYTHE_KYTHE_FACTS_EXTRACTOR_H_
+
+#include <string>
+#include <utility>
+
+#include "verilog/tools/kythe/indexing_facts_tree.h"
+#include "verilog/tools/kythe/indexing_facts_tree_extractor.h"
+
+namespace verilog {
+namespace kythe {
+
+// Node vector name for kythe facts.
+struct VName {
+  explicit VName(absl::string_view path, absl::string_view signature = "",
+                 absl::string_view root = "",
+                 absl::string_view language = "verilog",
+                 absl::string_view corpus = "https://github.com/google/verible")
+      : signature(absl::Base64Escape(signature)),
+        path(path),
+        language(language),
+        corpus(corpus),
+        root(root) {}
+
+  std::string ToString() const;
+
+  // Unique identifier for this VName.
+  std::string signature;
+
+  // Path for the file the name is extracted from.
+  std::string path;
+
+  // The language this name belongs to.
+  std::string language;
+
+  // The corpus of source code this VName belongs to
+  std::string corpus;
+
+  // A directory path or project identifier inside the Corpus.
+  std::string root;
+};
+
+// Facts for kythe.
+struct Fact {
+  Fact(const VName& vname, absl::string_view name, absl::string_view value)
+      : node_vname(vname),
+        fact_name(name),
+        fact_value(absl::Base64Escape(value)) {}
+
+  // The vname of the node this fact is about.
+  const VName& node_vname;
```

Comment:
It should be apparent from this being a `const&` member, but comment about the expected lifetime of this class with respect to other classes it "points-to".

Also be very cautious about pointer/iterator stability.
Any potentially re-allocating operation like a `vector::push_back` could potentially invalidate many pointers.
These considerations would be good to document in the high-level class comments -- explain what makes your design safe.

If you can narrow the interface that is exposed in public headers, that reduces chances and possibilities of incorrect use.  Are `Facts` and `Edge` important for the user of this class to know about, or could they be hidden as implementation details?

Reply:
How can know the expected life time of this class?
I don't know what u mean by safe design.
Fact and Edge are moved to other .h file.

Could u provide some more pointer about what y mean by safe-design and life time?

Reply:
For example, if a reference points to a vector element, any push_back on that vector could potentially move all of its elements to a new memory buffer thereby invalidating the old reference.